### ISSUE TITLE
Audio latency will keep accumulating, causing lag and inconsistency

### ISSUE THREAD

**Issue by AutumnSun1996** [rustdesk/rustdesk#534]
**Describe the bug you encountered:**

Found a large lag for Audio latency will keep accumulating, causing lag and inconsistency

**What did you expect to happen instead?**

Audio and video keep in sync.

**How did you install `RustDesk`?**

Both Windows portable version, from github releases.

---

**RustDesk version and environment**
Both RustDesk 1.1.8 on Windows10

The lag seems to keep accumulating, so after a long connection with audio on, there will be a noticeable lag.

I did a test with this video: https://www.bilibili.com/video/BV1m44y1t7Wh/
There is a lag for several seconds.

**Comment by rustdesk** [comment]
Try out 1.1.9 please.

**Comment by AutumnSun1996** [comment]
Will try 1.1.9 when I can access the remote machine again. just failed to install the new version remotely.
Is there any change for audio lag in 1.1.9?

**Comment by rustdesk** [comment]
> Will try 1.1.9 when I can access the remote machine again. just failed to install the new version remotely. Is there any change for audio lag in 1.1.9?

A lot

**Comment by AutumnSun1996** [comment]
The bug still exists in 1.1.9.
And I confirmed the steps to reproduce:
1. Connect to remote.
2. Turn audio on, and let something play on remote.
3. Wait for several minutes and the bug appears

Here is a video I recorded after 10 minutes wait, there is a lag of roughly 3 seconds.
[rustdesk-lag.zip](https://github.com/rustdesk/rustdesk/files/8699648/rustdesk-lag.zip)

**Comment by rustdesk** [comment]
@Heap-Hop Have a look please.

**Comment by Heap-Hop** [comment]
@AutumnSun1996 thx!

**Comment by rustdesk** [comment]
@AutumnSun1996 could you please try out https://github.com/rustdesk/rustdesk/releases/download/tmp/rustdesk-1.1.9-putes.exe? need reinstall on both side.

**Comment by AutumnSun1996** [comment]
Problem still exists.
Looked at the code #560, and seems u added some sort of timestamp at send time.
How about just increase playback rate if audio buffer at the receiver is large?

this is what i use for my websocket based remote desktop solution, and it works well

**Comment by rustdesk** [comment]
close to favor https://github.com/rustdesk/rustdesk/issues/3748

**Comment by rustdesk** [comment]
> this is what i use for my websocket based remote desktop solution, and it works well

Do you mean use a sample rate (playback rate) higher than real data sample rate? How do you caculate the playback rate to be used?

**Comment by rustdesk** [comment]
> > this is what i use for my websocket based remote desktop solution, and it works well
> 
> Do you mean use a sample rate (playback rate) higher than real data sample rate? How do you caculate the playback rate to be used?

@AutumnSun1996

**Comment by awsms** [comment]
I'm still getting this bug, even on direct local peer-to-peer connection

**Comment by rustdesk** [comment]
https://www.reddit.com/r/rustdesk/comments/1e4dis2/audio_is_not_in_sync_with_video_lags_behind_using/

**Comment by AutumnSun1996** [comment]
> > > this is what i use for my websocket based remote desktop solution, and it works well
> > 
> > 
> > Do you mean use a sample rate (playback rate) higher than real data sample rate? How do you caculate the playback rate to be used?
> 
> @AutumnSun1996

Sorry for just noticed it.
In my solution there is no playback rate, the viewer display the frames as soon as they are received and decoded.

**Comment by rustdesk** [comment]
That's because you are using a high-level API (browser help you handle the complexity)

**Comment by rustdesk** [comment]
Close as fix merged, please test the nightly build.

**Comment by rayangl** [comment]
1.3.1 PC端声音卡顿，不论PC远程移动端还是移动远程PC,声音都特别卡

**Comment by rayangl** [comment]
> 关闭修复合并后，请测试 nightly 构建（只需要更新控制端）。

1.3.1 PC端声音卡顿，不论PC远程移动端还是移动远程PC，声音都特别卡

**Comment by 21pages** [comment]
@rayangl could you please test the program linked here: https://github.com/21pages/rustdesk/releases/download/nightly/rustdesk-1.3.1-x86_64-find-proper-audio-buffer-length.exe? We're looking to determine the minimum proper audio buffer length. Currently, it’s set to 150 ms, but you can adjust it from the local side. Your findings will be helpful for us.

![1727766877121](https://github.com/user-attachments/assets/7a2cdc94-b8a4-43fd-ad06-021de62e0d3f)

**Comment by rayangl** [comment]
> 您能否测试一下此处链接的程序：https://github.com/21pages/rustdesk/releases/download/nightly/rustdesk-1.3.1-x86_64-find-proper-audio-buffer-length.exe？我们正在寻求**确定最小适当的**音频缓冲区长度。目前设置为 150 毫秒，你可以从本地端调整，然后连接到远端。您的发现将对我们有所帮助。
> 
> ![1727766877121](https://private-user-images.githubusercontent.com/14891774/372358777-7a2cdc94-b8a4-43fd-ad06-021de62e0d3f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3Njc2ODMsIm5iZiI6MTcyNzc2NzM4MywicGF0aCI6Ii8xNDg5MTc3NC8zNzIzNTg3NzctN2EyY2RjOTQtYjhhNC00M2ZkLWFkMDYtMDIxZGU2MmUwZDNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAxVDA3MjMwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkYmRkNDMzNmYwMjVlMmFjZWFlNGUzMWQ5YzVhZWI5YzQ4NmE5OGVhMzUzOGE3NDc4MDI4MjM2NWNmYzBmNWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._jI4KqoDCsvW_rVgfHhdA79XYXsiilq8h-rH8ck2_TA)

![image](https://github.com/user-attachments/assets/f5add2f4-4a36-4a72-b23b-950c626aedbc)
350ms 轻微卡顿 400ms就不卡顿了

**Comment by deivid-mudafy** [comment]
Hello, good evening. I’ve noticed that the issue is still occurring.
My setup consists of a local client running macOS and a remote server running Ubuntu.

I would really appreciate it if you could provide a build similar to rustdesk-1.3.1-x86_64-find-proper-audio-buffer-length.exe, but for Mac (aarch64.dmg).

Best regards!

**Comment by ghost** [comment]
its still a bug

**Comment by ghost** [comment]
https://github.com/rustdesk/rustdesk/issues/6177#issuecomment-2185008717


same issue pipewire

**Comment by ghost** [comment]
> We're looking to determine the minimum proper audio buffer length. Currently, it’s set to 150 ms, you can adjust it from the local side, then connect to the remote side. Your findings will be helpful for us.

had to get a windows pc to test it 150 and anything under is stuttering, 200 and above delay 3 seconds

i dont think its connecting client side issue to work on in the first place

**Comment by rustdesk** [comment]
I will build a test version for you. @hurrysh Thanks,

**Comment by GoGit-IT** [comment]
After 2 days of hacking I can confirm it's still an issue.
My setup:
- latest Debian 12 & KDE plasma - Proxmox VM - using Spice & Pipewire.
- Latest nightly clients & local selfhosted server.

I managed to resolve the lag temporarily by forcing my VM to use pulseaudio only.
(However I'm unable to retain that configuration for long)
After system restart, Rustdesk restart or a fly bumping in my window pipewire returns, so the 3s delay to.

**Comment by rustdesk** [comment]
pipewire is on which side? controlling or controlled? or no matter which side, it will have problem?

**Comment by rustdesk** [comment]
> had to get a windows pc to test it 150 and anything under is stuttering, 200 and above delay 3 seconds
> 
> i dont think its connecting client side issue to work on in the first place

Did you mean latest Windows version still has lag on your machine? You seemed testing https://github.com/21pages/rustdesk/releases/download/nightly/rustdesk-1.3.1-x86_64-find-proper-audio-buffer-length.exe 

How about current 1.4.0 for Windows (controlling side)?

@hurrysh

**Comment by GoGit-IT** [comment]
Hi
- Controller:
`Windows 11(Bare metal)`
- Controlled:
`Debian 12 with pipewire`
I was testing the latest nightly builds(2) on both clients & 1.4.0 only on both to. I did not touched any older versions.

**Comment by 21pages** [comment]
@GoGit-IT  Could you try out https://github.com/21pages/rustdesk/releases/tag/audio_delay on the controlling side ?

**Comment by GoGit-IT** [comment]
> [@GoGit-IT](https://github.com/GoGit-IT) Could you try out https://github.com/21pages/rustdesk/releases/tag/audio_delay on the controlling side ?

Hii!! BIG Ty for your work!
- Sadly the version above is somewhat worse. The audio delay is kinda shorter; between 2 - 2.3s but the sound stream is choppy and really not usable.
I turned off variable bit rate, did reboots and different sound sources, the result was the same.

Meanwhile I started up a Window11 VM on the very same Proxmox host. (SPICE drives same...) 0, lag or delay using the latest 1.4.0.

**Comment by ghost** [comment]
> How about current 1.4.0 for Windows (controlling side)?

the issue with 1.4.0. is constant delay (3 seconds), i tried the one you sent (1.3.x) as well described the issues with it 

> pipewire is on which side? controlling or controlled? or no matter which side, it will have problem?

controlled (the one im connecting to) side

> Did you mean latest Windows version still has lag on your machine?

yes, clearly some issue with pipewire handling, anydesk doesnt have it

**Comment by 21pages** [comment]
@GoGit-IT Is your Debian 12 a virtual machine?

**Comment by dmt-hub** [comment]
> > How about current 1.4.0 for Windows (controlling side)?
> 
> the issue with 1.4.0. is constant delay (3 seconds), i tried the one you sent (1.3.x) as well described the issues with it
> 
> > pipewire is on which side? controlling or controlled? or no matter which side, it will have problem?
> 
> controlled (the one im connecting to) side
> 
> > Did you mean latest Windows version still has lag on your machine?
> 
> yes, clearly some issue with pipewire handling, anydesk doesnt have it

Had the same issue - and here's a fix->>

Start rustdesk from the terminal with this command:
PULSE_LATENCY_MSEC=10 rustdesk

This will reduce the audiobuffer (you can monitor it by typing pw-top in the terminal)

**Comment by GoGit-IT** [comment]
> [@GoGit-IT](https://github.com/GoGit-IT) Is your Debian 12 a virtual machine?

Yes.

**Comment by GoGit-IT** [comment]
> Had the same issue - and here's a fix->>
> 
> Start rustdesk from the terminal with this command: PULSE_LATENCY_MSEC=10 rustdesk
> 
> This will reduce the audiobuffer (you can monitor it by typing pw-top in the terminal)

```
~$ PULSE_LATENCY_MSEC=10 rustdesk
Try setting transparent
Cannot load libcuda.so.1
Cannot load libcuda.so.1
flutter: launch args: []
flutter: initializing FFI main
flutter: _appType:main,info1-id:b2dde850e52f4e06bfa8bdd8682f2841,info2-name:Debian GNU/Linux,dir:/root
flutter: _globalFFI init
flutter: registerEventHandler callback_query_onlines address book peer
flutter: registerEventHandler load_address_book_peers address book peer
flutter: registerEventHandler callback_query_onlines group peer
flutter: registerEventHandler load_group_peers group peer
flutter: registerEventHandler callback_query_onlines recent peer
flutter: registerEventHandler load_recent_peers recent peer
flutter: registerEventHandler callback_query_onlines fav peer
flutter: registerEventHandler load_fav_peers fav peer
flutter: registerEventHandler callback_query_onlines discovered peer
flutter: registerEventHandler load_lan_peers discovered peer
flutter: _globalFFI init end
flutter: registerEventHandler native_ui native_ui
flutter: registerEventHandler check_software_update_finish check_software_update_finish
flutter: no window position saved, ignoring position restoration
flutter: handled by uni links: false
flutter: [MultiWindowHandler] active window changed: {0}
/usr/bin/systemctl
Try setting transparent

(rustdesk:920097): Gtk-CRITICAL **: 01:30:02.952: gtk_main_quit: assertion 'main_loops != NULL' failed
~$
```

TY for the hint!
Tomorrow I will do my best to keep Rustdesk running.
It's 2am here and when I start rustdesk so, the service it self is not running.
When I click to start it from the bottom status bar... does not end well.

**Comment by dmt-hub** [comment]
@GoGit-IT 
Don't know if it matters but I'm using this command on arch (: btw ;) - and not in a VM.
Do you have an nvidia card? there seems to be an issue with the cuda libraries that are missing "Cannot load libcuda.so.1"

The GTK error might indicate another RustDesk instance is running. Try:  pkill rustdesk

**Comment by GoGit-IT** [comment]
> [@GoGit-IT](https://github.com/GoGit-IT) Don't know if it matters but I'm using this command on arch (: btw ;) - and not in a VM. Do you have an nvidia card? there seems to be an issue with the cuda libraries that are missing "Cannot load libcuda.so.1"
> 
> The GTK error might indicate another RustDesk instance is running. Try: pkill rustdesk

Ty!
 - I'm using Debian.
 - The missing cuda libraries might be because hardware acceleration is enabled in Rustdesk. I will try without that soon, because there is no card attached to that VM, but the Spice drivers.
 - Before I run the command I made sure Rustdesk is terminated.
*Closed all taskbar presence and checked in btop*

**Comment by dmt-hub** [comment]
@GoGit-IT 
When running rustdesk the normal way - can you run pw-top in your terminal and check the value for Rustdesk under QUANT (audio buffer size). Here it was around 6000 (huge buffer - huge delay), and when running Rustdesk with the latency command it went to 480.

**Comment by GoGit-IT** [comment]
![Image](https://github.com/user-attachments/assets/7b331aef-fef5-4b20-88a1-08a20669d177)
Yes, I definitely see massive buffers set by default.
Running Rustdesk on the controlled side with `PULSE_LATENCY_MSEC=10 rustdesk` is REALY magnificent!
![Image](https://github.com/user-attachments/assets/5d0d0951-2e3c-49d7-928a-7213cebec59b)
- For some reason no Gtk-CRITICAL error neither today. Even tough I did NOT restart the vPC since.

The audio stream is really usable. Sure not perfect but a 40-50ms hiccup every 40 - 60seconds is truly acceptable and might be the result of many-many layers on my config.

Please implement this function to be configured/set and retained between system restarts.
TYtyTY for the awesome support!

**Comment by ghost** [comment]
![Image](https://github.com/user-attachments/assets/125b73aa-00f2-4b23-87cd-df4bb7e8a1a5)

clearly rustdesk issue doesnt set proper audio interface settings

**Comment by ghost** [comment]
as @dmt-hub mentioned its just misconfiguration of applications that run, same used to happen with chrome, 
add ie `PULSE_LATENCY_MSEC=83 PIPEWIRE_LATENCY=1024/48000` before rustdesk in service files and restart them after if u run it not as service just set these environment variables in bash/terminal before you run or even 256/48000 and MSEC to 5

more info at https://forum.manjaro.org/t/howto-troubleshoot-crackling-in-pipewire/82442

**Comment by ghost** [comment]
still a question why is rust desk using FLOAT 32 instead of signed int 16?

**Comment by YuChuXi** [comment]
> as [@dmt-hub](https://github.com/dmt-hub) mentioned its just misconfiguration of applications that run, same used to happen with chrome, add ie `PULSE_LATENCY_MSEC=83 PIPEWIRE_LATENCY=1024/48000` before rustdesk in service files and restart them after if u run it not as service just set these environment variables in bash/terminal before you run or even 256/48000 and MSEC to 5
> 
> more info at https://forum.manjaro.org/t/howto-troubleshoot-crackling-in-pipewire/82442

This works well for me. I've already written it to `/etc/systemd/system/rustdesk.service.d/override.conf` to automatically take effect.
```
[Service]
Environment="PULSE_LATENCY_MSEC=83"
```

**Comment by lukehb** [comment]
(Relevant for remote Linux machines only)

I was also experiencing an audio delay (2-3s) using RustDesk to remote into Ubuntu 24.04 machine that was using RustDesk client out of the box.

I resolved the audio delay by doing the following:

1. Open the terminal, enter `sudo nano /usr/lib/systemd/system/rustdesk.service`
2. Under the [Service] section add of the service file add: `Environment="PULSE_LATENCY_MSEC=60" "PIPEWIRE_LATENCY=1024/48000"` to add the relevant environment variables to configure pulse audio and pipewire.
3.  `Ctrl + O` + `Ctrl + X` to write and save the file
4. Restart the RustDesk service by typing the following commands in the terminal, first, `sudo systemctl daemon-reload` and then `sudo systemctl restart rustdesk.service` (note: this will disconnect any active RuskDesk session).

@rustdesk I think adding these environment variables to the service file created by the RustDesk install would be beneficial to many users. If this task is picked up in a separate issue please ping here so it can be followed. I have put together a PR here: https://github.com/rustdesk/rustdesk/pull/12724

**Comment by garbb** [comment]
I am also experiencing this same issue where the audio delay will gradually increase over the course of a session until it is around 1-2 sec behind the video.
This is not really consistent unfortunately; sometimes a session can go an hour without any delay and sometimes it can end up with a 1-2 sec delay after only 10-15mins.
I have tried all of the suggestions in this thread including 
`PULSE_LATENCY_MSEC=10`
and
`PULSE_LATENCY_MSEC=60
PIPEWIRE_LATENCY=1024/48000`
but unfortunately none have helped.
I am using the latest stable release 1.42 and a self-hosted server on windows 10 as the controlled side and also windows 10 on the controlling side.
As a sort of workaround I have noticed that if you check and uncheck the mute box on the controlling end, it will reset the delay. This is faster than reconnecting at least.

**Comment by lazycodeman** [comment]
> I am also experiencing this same issue where the audio delay will gradually increase over the course of a session until it is around 1-2 sec behind the video. This is not really consistent unfortunately; sometimes a session can go an hour without any delay and sometimes it can end up with a 1-2 sec delay after only 10-15mins. I have tried all of the suggestions in this thread including `PULSE_LATENCY_MSEC=10` and `PULSE_LATENCY_MSEC=60 PIPEWIRE_LATENCY=1024/48000` but unfortunately none have helped. I am using the latest stable release 1.42 and a self-hosted server on windows 10 as the controlled side and also windows 10 on the controlling side. As a sort of workaround I have noticed that if you check and uncheck the mute box on the controlling end, it will reset the delay. This is faster than reconnecting at least.

I encountered a similar issue where the sound would get interrupted with a slight cracking noise in the background. I discovered that the cause was a **mismatch in the output sound format** between the host and client PCs. By setting both machines' output sound settings to the same **24 bit, 48000 Hz**, the sound now works flawlessly. I hope this information is helpful.

_Edit:_ After further testing, I discovered that the output sound format of my headphones (client side) must match the Realtek(R) Audio format (client side). 

Make sure to check. Good luck!

**Comment by jfoxwoosh** [comment]
Happy to report the environment variables setting fixed the audio delay on Linux Mint 22.2.
I used Flatseal to set the variables on the host.

Host: Linux Mint 22.2 (based on ubuntu 24.04 LTS)
RustDesk Server:  flatpak v1.4.3

**Comment by Gwindalmir** [comment]
I have this issue when connecting to a Windows system from the web client, but I only notice it from the web client.

The latency is noticeable after only a couple of minutes and grows longer the longer the session is active. I've had the latency at least as high as 30 seconds. Noticed because of watching YouTube over the session.
Disconnecting and reconnecting resets the latency.

It happens with every codec.
